### PR TITLE
Allow any manifestly constant-evaluated expression and conversion in device function

### PR DESCRIPTION
**Pull request by keryell**
This should address https://github.com/KhronosGroup/SYCL-Docs/issues/267

### CODE REVIEW

**Comment by TApplencourt** [comment]
Look good enough to get us started :). In theory, if recursion is tail-recurvice if you be allowed to (as it will not need a call stack on the GPU)

**Review by TApplencourt — APPROVED** [review]

**Comment by keryell** [comment]
> Look good enough to get us started :). In theory, if recursion is tail-recurvice if you be allowed to (as it will not need a call stack on the GPU)

You assume that the compiler is smart enough to replace a tail-recursion by a loop. What if `-O0`? It looks like a QoI problem to be included in the specification.

**Comment by Naghasan** [comment]
I agree with Ronan, `What if -O0?` and what happens in debug as well (you may not want optimization there as well)

**Comment by keryell** [comment]
This has been generalized to address https://github.com/KhronosGroup/SYCL-Docs/issues/379

**Comment by fraggamuffin** [comment]
bug fix or NEXT or KHR?
NEXT or KHR?

**Comment by keryell** [comment]
Changing to a draft to wait for SYCL Next.

**Comment by tomdeakin** [comment]
Please open MR on GitLab. Thanks!

**Comment by keryell** [comment]
I have rebased this PR for SYCL Next as internal https://gitlab.khronos.org/sycl/Specification/-/merge_requests/740

**Comment by keryell** [comment]
Just removing the Draft status to consider this to be closed during next meeting and look at GitLab..

**Comment by gmlueck** [comment]
Recording our decision from the WG meeting on September 26, 2024:

* We should re-target this PR as a "big fix" to SYCL 2020
* @keryell volunteered to update the PR

File: adoc/chapters/device_compiler.adoc [hunk 1]
```diff
@@ -437,6 +446,16 @@ the entire device image when [code]#KernelA# is submitted to device
...
 
+[NOTE]
+====
+As stated in <<sec:language.restrictions.kernels>>, the restrictions
+for optional kernel features do not apply to manifestly
+constant-evaluated expressions or conversions in device code. Device
+code may use optional features in manifestly constant-evaluated
+expressions or conversions even if the device does not support the
+optional feature.
+====
```
Comment (gmlueck) [inline adoc/chapters/device_compiler.adoc:457]: I wonder if this should be normative text instead of a non-normative note.

**Review by gmlueck — COMMENTED** [review]
I think this looks good.  Just one small comment below.

File: adoc/chapters/device_compiler.adoc (hunk 1 above)
Comment (keryell) [reply adoc/chapters/device_compiler.adoc]: Done.

File: adoc/chapters/device_compiler.adoc [hunk 2]
```diff
@@ -377,6 +386,14 @@ in <<sec:backends>> must be defined by all conformant implementations.
...
+
+As stated in <<sec:language.restrictions.kernels>>, the restrictions for
+optional kernel features do not apply to manifestly constant-evaluated
+expressions or conversions in device code.
+Device code may use optional features in manifestly constant-evaluated
+expressions or conversions even if the device does not support the optional
+feature.
+
 As described in <<sec:device-aspects>>, an application can test whether a device
 supports these features by testing whether the device has an associated aspect.
```
Comment (gmlueck) [inline adoc/chapters/device_compiler.adoc:398]: Thanks for making the text normative, but I have a new comment now that the location of this paragraph has changed.  The phrase "these features" refers to the optional features described in the first paragraph.  However, your new paragraph intervenes, making it less clear what "these" refers to.  Let's change it to "an optional feature" to clarify what we are talking about:

```suggestion
supports an optional feature by testing whether the device has an associated aspect.
```

**Review by gmlueck — APPROVED** [review]

**Review by TApplencourt — APPROVED** [review]

**Comment by tomdeakin** [comment]
Please add "manifestly constant-evaluated expression" as a Glossary term with definition saying it is as defined in the C++20 specification.

**Comment by tomdeakin** [comment]
WG approved to merge to SYCL 2020

File: adoc/chapters/glossary.adoc [hunk 3]
```diff
@@ -375,6 +380,13 @@ For the full description please refer to <<subsec:buffers>>.
     In the SYCL interface an <<nd-range>> is represented by the [code]#nd_range#
     class (see <<subsubsec:nd-range-class>>).
 
+[[manifestly-constant-evaluated]]manifestly constant-evaluated expression or conversion::
+    ISO C++ +[expr.const]+ describes manifestly constant-evaluated expression or
+    conversion like constant expressions, condition of an +if constexpr+, an
+    immediate invocation, used in template parameters, in constant
+    initialization, etc. This is evaluated at compile-time by the compiler. For
+    more context, see <<sec:language.restrictions.kernels>>.
+
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply adoc/chapters/glossary.adoc]: Fixed in df3b0fa635a8ceb4d1673b4ae74a8bc43ed0d612

**Pull request merged**